datatree/datatree.py

```diff
@@ -217,10 +355,13 @@ def parent(self: DataTree, new_parent: DataTree) -> None:
         self._set_parent(new_parent, self.name)
 
     @property
-    def ds(self) -> Dataset:
-        """The data in this node, returned as a Dataset."""
-        # TODO change this to return only an immutable view onto this node's data (see GH #80)
-        return self.to_dataset()
+    def ds(self) -> DatasetView:
+        """
+        An immutable Dataset-like view onto the data in this node.
+        If you want a mutable Dataset containing the same data as in this node,
+        use `.to_dataset()` instead.
+        """
+        return DatasetView._from_node(self)
```

Comment:
@TomNicholas Why do you need to define a new `DatasetView` class inherited from `Dataset`?
Have you considered using `xarray.core.utils.Frozen`?
I.e., just returning `Frozen(self.to_dataset())` would not do the job?

cc: @aurghs

Reply:
I tried that, and it does give immutability nicely, but there are at least 2 other problems:

1) The `Frozen` object doesn't propagate methods, so you can't do `dt.ds.mean()` for example,
2) You would need to override the `__getitem__` method in order to implement out-of-node access like I tried to in #102.

We could define our own special `FrozenDataset` class which wraps `Dataset` instead of subclasses it though - that might be easier.

Reply:
Hi Tom,
concerning the second point, what would be a real use case for #102? Wouldn't it be better to force the user to use the datatree for out-of-node access, without modifying the dataset object?
Probably I'm missing something.

Reply:
Yes, I'm also not fully convinced about the need of out-of-node access for `.ds`. Sorry I didn't realise that in #80.
In general, I think that it's easier if `.ds` returns an object as close as possible to `xr.Dataset` (i.e., same object but immutable)

Reply:
> Maybe a 5th option could be to change Frozen in xarray and allow to pass on methods. 

Whether we change xarray's version of `Frozen` or make our own `FrozenDataset` class doesn't really matter, so that's basically the same as 3. I actually just tried that just now though by doing this

```python
class FrozenDataset(Frozen):
    """
    An immutable Dataset-like view onto the data in a single DataTree node.

    In-place operations modifying this object should raise an AttributeError.

    Operations returning a new result will return a new xarray.Dataset object.
    This includes all API on Dataset, which will be inherited.
    """

    FORBIDDEN_INPLACE_METHODS = ["update", "__setitem__"]

    def __getattr__(self, attr):
        if attr in dir(self.mapping) and attr not in self.FORBIDDEN_INPLACE_METHODS:
            return getattr(self.mapping, attr)
        else:
            raise AttributeError("%r object has no attribute %r" %
                                 (self.__class__.__name__, attr))
```
Unfortunately this doesn't work in that it doesn't have all the double-underscore arithmetic methods that Dataset has, so this happens in some of the tests:

```python
    @map_over_subtree
    def times_ten(ds):
>       return 10.0 * ds
E       TypeError: unsupported operand type(s) for *: 'float' and 'FrozenDataset'
```

Reply:
> I took a look and I think (4) would be the best option. Maybe the easiest thing to do is implementing (1) for now, and switch to https://github.com/pydata/xarray/pull/6086 when it's ready.

I'm not sure if we'll need (4), but I agree with implementing (1) for now definitely.

> Question to better understand the overall scope of .ds: As it doesn't look straightforward to have an immutable .ds, what's the advantage of having both to_dataset() and .ds.

If we have (4) then we *might* not need both. But the general idea was that `.ds` returns a view onto the data in that node, and is "connected" to he contents in the node, whereas `.to_dataset()` returns a completely new and unconnected object.

> If it's just an API design, maybe it's easier to maintain and document a single method (e.g., something like dt.get_ds(node_name: None | str))

I'm not a big fan of that idea because we are implementing `__getitem__` to select a node, and `dt[node_name].ds` is already slightly shorter (and IMO more intuitive) than `dt.get_ds(node_name)`. We could have multiple methods though, I don't think it's particularly hard to maintain multiple methods of data access if they are useful.

Reply:
Makes sense - thanks!
I'd say go ahead and merge this and #41 when you're ready so you can start getting more feedback from users.